include/oneapi/dpl/pstl/algorithm_ranges_impl.h

```diff
@@ -0,0 +1,315 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_ALGORITHM_RANGES_IMPL_H
+#define _ONEDPL_ALGORITHM_RANGES_IMPL_H
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+
+#include <ranges>
+#include "algorithm_fwd.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+namespace __ranges
+{
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_for_each
+//---------------------------------------------------------------------------------------------------------------------
+
+template <typename _Tag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each_impl(_Tag __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    static_assert(__is_parallel_tag_v<_Tag> || typename _Tag::__is_vector{});
+
+    auto __f_1 = [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, __val));};
```

Comment:
Should we forward `__val` here since it's a forwarding reference and in other similar places?

Reply:
yeah, you are right, thank you